lua/ttt2/libraries/door.lua

```diff
@@ -0,0 +1,589 @@
+
+		-- set up synced states if on server
+		if CLIENT then continue end
+
+		ent:SetNWBool("ttt2_door_locked", ent:GetInternalVariable("m_bLocked") or false)
+		ent:SetNWBool("ttt2_door_forceclosed", ent:GetInternalVariable("forceclosed") or false)
+		ent:SetNWBool("ttt2_door_open", door.IsOpen(ent) or false)
+
+		ent:SetNWBool("ttt2_door_player_use", door.PlayerCanUse(ent))
+		ent:SetNWBool("ttt2_door_player_touch", door.PlayerCanTouch(ent))
+		ent:SetNWBool("ttt2_door_auto_close", door.AutoCloses(ent))
+		ent:SetNWBool("ttt2_door_is_destructable", door.IsDestructible(ent))
+
+		entityOutputs.RegisterMapEntityOutput(ent, "OnOpen", "TTT2DoorOpens")
+		entityOutputs.RegisterMapEntityOutput(ent, "OnClose", "TTT2DoorCloses")
+		entityOutputs.RegisterMapEntityOutput(ent, "OnFullyOpen", "TTT2DoorFullyOpen")
+		entityOutputs.RegisterMapEntityOutput(ent, "OnFullyClosed", "TTT2DoorFullyClosed")
+
+		HandleDoorPairs(ent)
+
+		if cvDestructableDoor:GetBool() and not (cvDestructableDoorLocked:GetBool() and ent:IsDoorLocked()) then
+			ent:MakeDoorDestructable(true)
+		end
+	end
+
+	door_list.doors = doors
+end
+
+---
+-- Returns all valid door entity class names
+-- @return table A table of door class names
+-- @realm shared
+function door.GetValid()
+	return valid_doors
+end
+
+---
+-- Returns if a passed door class is a valid normal door (prop_door_rotating)
+-- @return boolean True if it is a valid normal door
+-- @realm shared
+function door.IsValidNormal(cls)
+	return valid_doors.normal[cls] or false
+end
+
+---
+-- Returns if a passed door class is a valid special door (func_door, func_door_rotating)
+-- @return boolean True if it is a valid special door
+-- @realm shared
+function door.IsValidSpecial(cls)
+	return valid_doors.special[cls] or false
+end
+
+---
+-- Returns all valid door entities found on a map
+-- @return table A table of door entities
+-- @realm shared
+function door.GetAll()
+	return door_list.doors
+end
+
+if SERVER then
+	---
+	-- Returns if a door is open by reading out the internal variable.
+	-- @param Entity ent The door entity that should be checked
+	-- @return boolean Returns if the door is open
+	-- @internal
+	-- @realm server
+	function door.IsOpen(ent)
+		local cls = ent:GetClass()
+
+		if door.IsValidNormal(cls) then
+			return ent:GetInternalVariable("m_eDoorState") ~= 0
+		elseif door.IsValidSpecial(cls) then
+			return ent:GetInternalVariable("m_toggle_state") == 0
+		end
+
+		return false
+	end
+
+	---
+	-- Returns if a player can use interact with a door by reading out the internal variable.
+	-- @param Entity ent The door entity that should be checked
+	-- @return boolean Returns if the door reacts to a player use
+	-- @internal
+	-- @realm server
+	function door.PlayerCanUse(ent)
+		local cls = ent:GetClass()
+
+		if door.IsValidNormal(cls) then
+			-- 32768: ignore player +use
+			return not ent:HasSpawnFlags(32768)
+		elseif door.IsValidSpecial(cls) then
+			-- 256: use opens
+			return ent:HasSpawnFlags(256)
```

Comment:
Why not using the globals directly? ref.: https://wiki.facepunch.com/gmod/Enums/SF

Reply:
because these are different enums. Each entity has their own set of numbers. 

ref1: https://developer.valvesoftware.com/wiki/Prop_door_rotating
ref2: https://developer.valvesoftware.com/wiki/Func_door

As you can see, not even these two types of doors have the same numbers for the same features.

Reply:
You are right, but I mean why not creating matching globals? :D

Reply:
I though about this. However I don't know if anyone would need them and I had to make two sets of them for the two types of doors. If you'd prefer this, I could add it.

Reply:
I personally would like to have them. This would make the door module more professional (same as LeBroomer did with his thermal vision). Anyways, just if you want :)

Reply:
I'll add them!